File: backend/pkg/database/models/project.go
package models

import (
	"time"

	"blog-agent-go/backend/internal/core/project"

	"github.com/google/uuid"
	"github.com/lib/pq"
)

// ProjectModel is the GORM model for projects
type ProjectModel struct {
	ID          uuid.UUID     `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Title       string        `json:"title" gorm:"not null"`
	Description string        `json:"description" gorm:"type:text;not null"`
	Content     string        `json:"content" gorm:"type:text"`
	TagIDs      pq.Int64Array `json:"tag_ids" gorm:"type:integer[]"`
	ImageURL    string        `json:"image_url"`
	URL         string        `json:"url"`
	CreatedAt   time.Time     `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt   time.Time     `json:"updated_at" gorm:"autoUpdateTime"`
}

func (ProjectModel) TableName() string {
	return "project"
}

// ToCore converts the GORM model to the domain type
func (m *ProjectModel) ToCore() *project.Project {
	return &project.Project{
		ID:          m.ID,
		Title:       m.Title,
		Description: m.Description,
		Content:     m.Content,
		TagIDs:      m.TagIDs,
		ImageURL:    m.ImageURL,
		URL:         m.URL,
		CreatedAt:   m.CreatedAt,
		UpdatedAt:   m.UpdatedAt,
	}
}

// ProjectModelFromCore creates a GORM model from the domain type
func ProjectModelFromCore(p *project.Project) *ProjectModel {
	return &ProjectModel{
		ID:          p.ID,
		Title:       p.Title,
		Description: p.Description,
		Content:     p.Content,
		TagIDs:      p.TagIDs,
		ImageURL:    p.ImageURL,
		URL:         p.URL,
		CreatedAt:   p.CreatedAt,
		UpdatedAt:   p.UpdatedAt,
	}
}

File: backend/src/database/models/project.rs
use chrono::{DateTime, Utc};
use diesel::{Identifiable, Insertable, Queryable, Selectable};
use uuid::Uuid;

use crate::{core::project::Project, error::AppError, schema::project};

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = project)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct ProjectRow {
    pub id: Uuid,
    pub title: String,
    pub description: String,
    pub image_url: Option<String>,
    pub url: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub content: Option<String>,
    pub tag_ids: Option<Vec<Option<i32>>>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = project)]
pub struct NewProjectRow<'a> {
    pub id: Uuid,
    pub title: &'a str,
    pub description: &'a str,
    pub image_url: &'a str,
    pub url: &'a str,
    pub content: &'a str,
    pub tag_ids: Vec<Option<i32>>,
}

impl TryFrom<ProjectRow> for Project {
    type Error = AppError;

    fn try_from(row: ProjectRow) -> Result<Self, Self::Error> {
        let tag_ids = row
            .tag_ids
            .unwrap_or_default()
            .into_iter()
            .map(|id| id.map(i64::from).ok_or(AppError::Database))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Self {
            id: row.id,
            title: row.title,
            description: row.description,
            content: row.content.unwrap_or_default(),
            tag_ids,
            image_url: row.image_url.unwrap_or_default(),
            url: row.url.unwrap_or_default(),
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}
